cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -0,0 +1,307 @@
+                                            int fetcherPersistThreshold) {
+    super(metricDatasetFactory);
+    this.metricsTopics = new TopicId[partitions.size()];
+    this.metricsTopicMetaKeys = new TopicIdMetaKey[partitions.size()];
+    for (int i = 0; i < partitions.size(); i++) {
+      this.metricsTopics[i] = NamespaceId.SYSTEM.topic(topicPrefix + "_" + i);
+      this.metricsTopicMetaKeys[i] = new TopicIdMetaKey(this.metricsTopics[i]);
+    }
+    this.messagingService = messagingService;
+    try {
+      this.recordSchema = schemaGenerator.generate(MetricValues.class);
+      this.recordReader = readerFactory.create(TypeToken.of(MetricValues.class), recordSchema);
+    } catch (UnsupportedTypeException e) {
+      throw Throwables.propagate(e);
+    }
+    this.metricStore = metricStore;
+    this.fetcherPersistThreshold = fetcherPersistThreshold;
+    this.metricsContextMap = Collections.<String, String>emptyMap();
+    processMetricsThreadList = new ArrayList<>();
+  }
+
+  @Override
+  public void setMetricsContext(MetricsContext metricsContext) {
+    this.metricsContext = metricsContext;
+    this.metricsContextMap = metricsContext.getTags();
+    metricStore.setMetricsContext(metricsContext);
+  }
+
+  @Override
+  protected void run() {
+    LOG.info("Start running MessagingMetricsProcessorService");
+    MetricsConsumerMetaTable metaTable = getMetaTable();
+    if (metaTable == null) {
+      LOG.info("Could not get MetricsConsumerMetaTable, seems like we are being shut down");
+      return;
+    }
+    int i = 0;
+    while (i < metricsTopics.length && isRunning()) {
+      byte[] messageId = null;
+      try {
+        messageId = metaTable.get(metricsTopicMetaKeys[i]);
+        LOG.info("Last processed MessageId for topic: {} is {}", metricsTopics[i], messageId);
+      } catch (Exception e) {
+        LOG.info(String.format("Cannot retrieve last processed MessageId for topic: %s", metricsTopics[i]), e);
+      }
+
+      MessageFetcher fetcher;
+      try {
+        fetcher = messagingService.prepareFetch(metricsTopics[i]);
+      } catch (Exception e) {
+        LOG.error(String.format("Failed to create fetcher for topic: %s", metricsTopics[i]), e);
+        return;
+      }
+      if (messageId == null) {
+        // If no messageId is found for the last processed message, start fetching from beginning
+        fetcher.setStartTime(0L);
+      } else {
+        fetcher.setStartMessage(messageId, false);
+      }
+      ProcessMetricsThread processMetricsThread = new ProcessMetricsThread(fetcher, metricsTopics[i],
+                                                                           metricsTopicMetaKeys[i]);
+      processMetricsThread.setDaemon(true);
+      processMetricsThread.start();
+      processMetricsThreadList.add(processMetricsThread);
+      i++;
+    }
+
+    for (ProcessMetricsThread thread : processMetricsThreadList) {
```

Comment:
It seems that with the current implementation, naming threads is more straightforward and code complexity is almost the same when starting and waiting for threads to finish?